src/client/datascience/kernel-launcher/localKernelFinder.ts

```diff
@@ -62,8 +65,22 @@ export class LocalKernelFinder implements ILocalKernelFinder {
                 ? PYTHON_LANGUAGE
                 : getTelemetrySafeLanguage(getLanguageInNotebookMetadata(notebookMetadata) || '');
         try {
+            const preferredKernelFromCache = this.findPreferredLocalKernelConnectionFromCache(notebookMetadata);
             // Get list of all of the specs
-            const kernels = await this.listKernels(resource, cancelToken, 'useCache');
+            const kernelsPromise = this.listKernels(resource, cancelToken, 'useCache');
+
+            // If we have a value from cache, return that.
+            // Check if this is still valid (i.e. the interpreter is still installed).
+            if (
+                preferredKernelFromCache &&
+                preferredKernelFromCache.interpreter &&
+                (await this.fs.localFileExists(preferredKernelFromCache.interpreter.path))
+            ) {
+                traceInfo(`Freferred kernel connection found in cache ${preferredKernelFromCache.id}`);
+                return preferredKernelFromCache;
```

Comment:
No need to search through all the kernels if we have a cached kernel that matches exactly with what's in the notebook metadata.